### ISSUE TITLE
real time input and output for training LSTM/RNN

### ISSUE THREAD

**Issue by xcszbdnl** [fchollet/keras#7317]
Hi, guys.

I'd like to train a recurrent network. However, they input and output orders are determined in real time. Also, they have different length for each sequence. For example, I have training data like this:

          y1 y2 y3 y4 y5
    x1 = [1, 0, 1]
    x2 = [0, 1, 0, 1]
    x3 = [1, 1, 0, 0, 1]
    x4 = [1, 0, 1, 1]
    ....

When I train a LSTM, the first input, target is determined by a function `determineFirstOrder()`, (e.g.`x1->(y1,y2) x2->(y2,y1), x3->(y4,y5), x4->(y1,y2)...`). Therefore, the input vector will be`1 1 0 1...`, target vector will be `0 0 1 0`

The next input and output is also determined by a function `determineNextOrder()`, but this will take the output of the LSTM into consideration. 

The question is:
1. How to deal with the variable length in the training data? (This may be done by mask layers. Padding the sequence is also an option, but in my case, the padding zeros will effect the LSTM state. Therefore, I'd like to use the mask layers.)

2. How to feed the real-time inputs and outputs into LSTM? The t-th inputs and outputs are determined by the output of the LSTM (at time t-1).

If I am missing something, please let me know I would really appreciate it!

Thanks!

**Comment by td2014** [comment]
@xcszbdnl .  Can you please clarify how your training procedure works?  I'm not quite following your notation x1 -> (y1,y2) etc.  Are the y1 and y2 related to the example training data you show?  Thanks.

**Comment by xcszbdnl** [comment]
@td2014 Yes, the input and output are related to every instance of training data. For x1-> (y1, y2), which means I should input the data of y1 to predict the data of y2. Then I will use `determineNextOrder()` to determine which y to predict. e.g. If y2 == 0, then I will use the data of y2 to predict y3, else if y2 == 1, then I will use the data of y2 to predict y4.  Things go on like this util all y are predicted. The final orders may like this y1 as input, and the order of (y2, y3) as output. Or it can be y1 as input, the order of (y3, y2) as output. The orders are determined by a function at each timestep.

**Comment by td2014** [comment]
@xcszbdnl .  Just to be clear, you are saying that you determine a new target at each time step with depends on the output of the RNN at the previous time step.  I'm just trying to make sure I understand your architecture.  So, just taking sample x1 you would have something like this:

x1(t=0)-->LSTM-->y(t=0) -->function-->target(t=1)
x1(t=1)-->LSTM-->y(t=1) -->function -->target(t=2)
...
x1(t=N)-->LSTM-->y(t=N)

Then at timestep t=N:
Compute loss for sample X1 and back propagate weight update, taking into account errors between LSTM output at each timestep and target at that timestep.  Is this correct?  Thanks.

**Comment by xcszbdnl** [comment]
@td2014 Let me describe it more details: If x1 are input features, and y is a vector of length t.
t = 0, (x1, 0)-->LSTM-->target y(column=1)-->function-->next target column should be 2
t = 1, (x1, y1)-->LSTM-->target y(column=2)-->function-->next target column should be 4
t = 2, (x1, y2)-->LSTM-->target y(column = 4)-->function-->next target column should be 3
....

 I will use x1 and y as input features to LSTM. At timestep t, which column to predict is determined by a function `determineNextOrder(previous_output)`, previous_output is the output of LSTM at timestep t-1.
Is it more clearly?
